src/mailroom_madness.py

```diff
@@ -0,0 +1,267 @@
+    while True:
+        user_input = input(prompt)
+        if user_input == u'cancel':
+            return False
+        elif validate_donation_amount(user_input):
+            return int(user_input)
+        else:
+            out.write(u'{}\n'.format(failure))
+
+
+def validate_donation_amount(user_input):
+    """Take user input, returns boolean."""
+    try:
+        return int(user_input) > 0
+    except ValueError:
+        return False
+
+
+def send_thank_you(donor_data):
+    """Prompt user for a name and donation amount and records them.
+
+    Args:
+        donor_data (dict)
+    """
+    name = prompt_donor_name(donor_data)
+    if name is False:
+        return
+
+    amount = get_donation_amount()
+    if amount is False:
+        return
+
+    log_thank_you(name, amount, donor_data)
+
+
+def log_thank_you(name, num, donor_data, out=sys.stdout):
+    """Record new donation and prints a report.
+
+    Args:
+        name (str)
+        num (int)
+        donor_data (dict)
+        out (file, optional)
+    """
+    donor_data[name].append(num)
+    message = u'Thanks {} for your generous donation of {} dollars.\n'
+    out.write(message.format(name, num))
+
+
+def align_cell(value, padding):
+    """Align a single value using padding.
+
+    Args:
+        value (str or int)
+        padding (int)
+    """
+    whitespace = u' ' * (padding - len(str(value)) + 2)
+    return u'{}{}'.format(value, whitespace)
+
+
+def max_size(t):
+    """Return largest value from t by its string representation.
+
+    Args:
+        t (list): A list of values to be process.
+    """
+    return max(map(lambda x: len(str(x)), t))
+
+
+def get_paddings(rows):
+    """Get the whitespace padding amount per column.
+
+    Args:
+        rows (list)
+    """
+    return [max_size(map(lambda row: row[i], rows))
+            for i in range(len(rows[0]))]
+
+
+def print_row(row, padding, out=sys.stdout):
+    """Print row to out with supplied padding.
+
+    Args:
+        row (tuple)
+        padding (int)
+        out (file, optional): The file to print to.
+
+    Returns:
+        TYPE: Description
+    """
+    for value, padding in zip(row, padding):
+        out.write(align_cell(value, padding))
+    out.write(u'\n')
+
+
+def print_table(rows, out=sys.stdout):
+    """Print an entire table to the console.
+
+    Args:
+        rows (list): The table as a list of rows
+        out (file, optional): The file to print to.
+
+    Returns:
+        TYPE: Description
+    """
+    paddings = get_paddings(rows)
+    for row in rows:
+        print_row(row, paddings, out=out)
+
+
+def generate_row(name, donations):
+    """Generate a single row from name and donations.
+
+    Args:
+        name (str): The name of the donor
+        donations (list): A list of donations
+
+    Returns:
+        tuple
+    """
+    total = sum(donations)
+    donation_count = len(donations)
+    average = total / (donation_count or 1)
```

Comment:
watch out for this division in Python2 vs. Python 3.  It won't work the same way in both places.